### ISSUE TITLE
Proposed plan on the next phase

### ISSUE THREAD

**Issue by lextm** [dotdevelop/dotdevelop#33]
As now #29 is done and we have a working build on Ubuntu 20.04 (with the old editor and without Git integration), the first phase of taking over MonoDevelop might be considered done,

* Grouped a few developers around the fork.
* Revived the old editor to fill in the gaps.
* Amazingly integrated with the 8.x branch so many fixes contributed earlier by Microsoft can still be used.

The next phase I think is to ship new releases to end users (as not everyone is supposed to build from the source code). That yields a list of possible tasks below,

## Supported Platform Expansion
Expand the CI platforms (right now Ubuntu 20.04) to a richer set (all major Linux distributions and their active LTS versions). That can be ambitious, but we can rule out certain less popular distributions or older versions if technical difficulty requires too much effort.

> We can create new issues (probably each tracks an OS version) and pull requests so that everyone wanting to participate can pick up one or two.

## Version Bump
Maybe we should create a new branch for future development and call it version 9.

## Taking Over More MonoDevelop Assets
Now the source code takeover is done, and we have a foundation to work on. I think next is to take over more assets from Microsoft in order to prepare a new release.

* Set a clear scope to ship MonoDevelop 9 only for Linux. Ignore Windows/macOS builds at least in the near future.
* Contact Microsoft to see if they can transfer the MonoDevelop homepage ownership, as well as Linux packaging infrastructure.
  * If such assets can be transferred in any suitable way, then MonoDevelop is revived directly and new releases can come smoothly. We improve the CI/CD pipeline to build/pack for supported platforms and ship to end users as version 7/8 to version 9 upgrade.
  * If the transfer is not possible, then time to find another suitable brand and recreate the packaging infrastructure.

## Notes
This plan focuses on packaging, branding, and possible Microsoft transfer. What it does not cover is all other features and bug fixes that the small group of developers can contribute in parallel.

We might ship some new features,

* Language Server Protocol support
* builtin .NET Core debugger
* GtkSharp3
* .NET 5 based MonoDevelop

and some bug fixes (like #30 ) to make a richer version 9 release, but even without such a new release based on the current code base can be a major upgrade for those users on version 7.

With more .NET Core features/bug fixes added, we can promote this tool as MonoDevelop for .NET Core/.NET 5/.NET 6, which is still easier than promoting a new brand.

**Comment by knocte** [comment]
> Expand the CI platforms (right now Ubuntu 20.04) to a richer set (all major Linux distributions and their active LTS versions).

I wholeheartedly disagree with this, I think it's a wasted effort, not to mention that GithubActionsCI doesn't support a Linux distro different than ubuntu.

My point is that the future is to be cross-distro via some packaging system like snap or flatpak. I have a mono app myself distributed via snap so I've already solved many hurdles about distributing a mono app this way, which may save us some time if I try to contribute something based on this knowledge.

> Ignore Windows/macOS builds at least in the near future.

But why? Look, I think Microsoft had a point about not supporting Linux because the effort to get the same level of quality for the majority of distros doesn't pay off, but I think only supporting Ubuntu would have not such a big deal for them. Now you want to go the complete opposite direction... for what reason? If we can manage to create win&macOS builds, I don't see why not try to keep the code crossplatform, unlike what Microsoft did.

> as well as Linux packaging infrastructure.

As I already hinted above, I don't think we need this if we just focus on snap or flatpak. Doing something different would be too much overhead for quite a low gain.

> Language Server Protocol support

With regards to this, I think it's wise for MS to converge their LSP story between VSCode and VStudio first, and after that try to refactor the CSharpBinding and FSharpBinding to try to reuse that work (so that MD gets up to par with VSCode at least).

> With more .NET Core features/bug fixes added, we can promote this tool as MonoDevelop for .NET Core/.NET 5/.NET 6, which is still easier than promoting a new brand.

I'd say let's not discuss this in here, as there is another github issue for that.

**Comment by lextm** [comment]
> I wholeheartedly disagree with this, I think it's a wasted effort, not to mention that GithubActionsCI doesn't support a Linux distro different than ubuntu.

GitHub Actions allows you to use any Docker image you like, so Ubuntu is clearly not the only OS you can use.

> My point is that the future is to be cross-distro via some packaging system like snap or flatpak. I have a mono app myself distributed via snap so I've already solved many hurdles about distributing a mono app this way, which may save us some time if I try to contribute something based on this knowledge.

The nightmare of flatpak was just a few years ago. I agree snap might be a good option to go, only if you have enough time to get it working. But if Microsoft can transfer the packaging infrastructure (or at least their packaging scripts) to this team, then shipping packages for different distributions should be quick.

> But why? Look, I think Microsoft had a point about not supporting Linux because the effort to get the same level of quality for the majority of distros doesn't pay off, but I think only supporting Ubuntu would have not been such a big deal for them. Now you want to go the complete opposite direction... for what reason? If we can manage to create win&macOS builds, I don't see why not try to keep the code crossplatform, unlike what Microsoft did.

Windows/macOS build is not trivial. You can keep a great goal of going cross platform, but a more realistic approach is to serve where you are needed the most.

> As I already hinted above, I don't think we need this if we just focus on snap or flatpak. Doing something different would be too much overhead for quite a low gain.

I don't see much overhead here. For example, you can focus on snap adoption, while I might work with Microsoft on the transfer.

> With regards to this, I think it's wise for MS to converge their LSP story between VSCode and VStudio first, and after that try to refactor the CSharpBinding and FSharpBinding to try to reuse that work (so that MD gets up to par with VSCode at least).

Microsoft probably won't share their code any more in that area (VS for Windows LSP libraries are not open sourced), while VS for Windows/Mac clearly use more and more LSP based components (XAML/Razor and more).

How to adopt LSP in MonoDevelop can be a long term project to work on, so how much might be delivered in "version 9" can be adjusted. I think the goal is to ship a new release as early as possible to maintain a large enough user base, and probably draw some new users.

> I'd say let's not discuss this in here, as there is another github issue for that.

Agree. Also it can be discussed once this team do have a new release ready to ship. Right now everything is subject to change.

**Comment by knocte** [comment]
> Microsoft probably won't share their code

If you were talking about VS here, I would agree. However, I mentioned VSCode, which is opensource. I don't think they would want to shoot themselves in the foot by turning their OmniSharp/Ionide plugins into proprietary. And AFAIU (I read it somewhere, don't remember where) there will be convergence between code of VSCode and VS(for Windows and for macOS).

**Comment by lextm** [comment]
> If you were talking about VS here, I would agree. However, I mentioned VSCode, which is opensource. I don't think they would want to shoot themselves in the foot by turning their OmniSharp/Ionide plugins into proprietary. And AFAIU (I read it somewhere, don't remember where) there will be convergence between code of VSCode and VS(for Windows and for macOS).

The code quality of OmniSharp can be tricky. For some of my open source projects, MonoDevelop 7 gives much better performance than VSCode/OmniSharp.

While VSCode/OmniSharp/Roslyn are all open sourced, you can find it hard to replicate the experience of VS for Windows. VS for Mac might be catching up, but again that's a closed source product now. It is hard to predict whether Microsoft is willing to share the knowledge behind for free, or JetBrains Rider will be the first to clone that.

**Comment by knocte** [comment]
> I won't try to predict whether Microsoft is willing to share the knowledge behind for free

I don't think they would be willing to do that. But if they converge their LSP effort with VSCode, they have no other choice.

> The code quality of OmniSharp can be tricky. For some of my open source projects, MonoDevelop 7 gives much better performance than VSCode/OmniSharp. While VSCode/OmniSharp/Roslyn are all open sourced, you can find it hard to replicate the experience of VS for Windows.

TBH I haven't played with OmniSharp much, so can't comment on that. However, I played a lot with MonoDevelop's F# binding (which was the same as VS4Mac's before it stopped being opensource) and it was very very low quality (and I have reasons to believe that VSCode's counterpart, Ionide, was and is much better).

**Comment by lytico** [comment]
some thoughts:

- about branding (MonoDevelop)

maybe we should try to be a project of https://dotnetfoundation.org, this could open some doors

- refactoring the GUI to GtkSharp 

(https://github.com/dotdevelop/GtkSharp) is the next critical step. Without that no dotnetcore / .net5-version
this is a huge task!

- using more Roslyn / LanguageServer 

with some adoptions OmniSharp (= seen as a Roslyn-Wrapper) 
to have recent language support

**Comment by lytico** [comment]
- about versioning

maybe change to semantic versioning, eg. 20.12 for https://github.com/dotdevelop/dotdevelop/tree/dotdevelop_oe_8.4.3.12

**Comment by lytico** [comment]
- get latest monobranch (https://github.com/mono/monodevelop/tree/master) working with Old Editor
 first draft is here: https://github.com/dotdevelop/dotdevelop/tree/dotdevelop_oe_8.6

**Comment by vmcbaptista** [comment]
Would be nice to organize the plan into a backlog, maybe using a GitHub Project Board.

**Comment by lytico** [comment]
we have one, but it's little bit orphaned:
https://github.com/orgs/dotdevelop/projects

**Comment by lextm** [comment]
I monitored this repo for a while and decided that the initial phase (code editor + .NET 5/6) is almost done. Close this issue now. Might open a new issue to suggest what might be considered in the next phase.

**Comment by hwthomas** [comment]
@lextm 
>Might open a new issue to suggest what might be considered in the next phase.  

I would welcome this.  Personally, I would like to canvas ideas as to how the codebase might be cleaned up as we progress.  

From what as I've seen while getting more familiar with several areas of DotDevelop, it appears that code has simply been duplicated into DotDevelop (for example, in the case of MonoDevelop.PackageManagement,  from NuGet.Client/Core), thus creating a parallel (and difficult to maintain) implementation rather than using the public API to NuGet.  The same is true of MSBuild.IntrinsicFunctions, etc.  

Cleaning up and enforcing the public interfaces between the DotDevelop IDE GUI, and the various integrated components, would make DotDevelop a much better project to work on, but it's no easy task!  

I look forward to your new Issue and road map.

**Comment by gluckez** [comment]
@hwthomas indeed, it would be a good idea to open a new issue to suggest next phases, but also to de-orphan the backlog here: https://github.com/orgs/dotdevelop/projects?query=is%3Aopen and draw more attention to that by adding it to the wiki.
This way, people can easily see the roadmap and it's a bit more encouraging to participate because they know work is being done on it.
